packages/sls-env/src/__tests__/dependencies.test.ts

```diff
@@ -174,5 +174,89 @@ describe('serverless environment', () => {
         new EventDependencyError("No event based dependency found for 'buildMessage'")
       )
     })
+
+    it('correctly resolves dependency factories dependent on event', async () => {
+      type BuildMessage = (message: string, name: string) => string
+      type BuildMessageDependencies = {
+        buildMessage: DependencyFactory<BuildMessage>
+      }
+
+      const buildMessagesFactory = () => ({ event }: EventPayload) => {
+        switch (event.message) {
+          case 'client1':
+            return (_m: string, _n: string) => `Using client1!`
+          case 'client2':
+            return (_m: string, _n: string) => `Using client2!`
+          default:
+            throw new Error(`Cannot resolve dependency based on event message: ${event.message}`)
+        }
+      }
+
+      const handler = environment<
+        Handler<MessageEvent, NameContext, string>,
+        any,
+        BuildMessageDependencies,
+        EventPayload
+      >()
+        .global({
+          buildMessage: setDependencyFactory<BuildMessage, EventPayload>(buildMessagesFactory())
```

Comment:
not sure about the name `setDependencyFactory` as that actually does not set but rather `creates` 

IMO `dependencyFactory` would be a better name

Reply:
potentially having following 

```ts
type FactoryDependency<T> = () => (e: EventPayload) => T

....

const createBuildMessage: FactoryDependency<BuildMessage> = ...
....

buildMessage: dependencyFactory(createBuildMessage) 

... 

```

Or alternatively, the factory has to return dependency and we can have:

```ts
type FactoryDependency<T> = () => (e: EventPayload) => DependencyFactory<T>
```

Reply:
* function name: yeah you are right.. I was playing with the names and eventually ended up to `setDependencyFactory` .. I did have `dependencyFactory` in the beginning but changed it later.

* the invocation of `buildMessagesFactory()`doesn't make sense much in the context of this test alone. In a real scenario the user woud pass in some parameters or config to the buildMessagesFactory() function which will then be used when the resolver function is invoked with the event/payload/context, and I just wanted to illustrate that. I will update the example to better show this. 
But yeah the main point is that we just need to return a function that will be passed the event/payload/context.

Reply:
The parameter passed to `setDependencyFactory` should be of type `DependencyFactoryResolver`

```
type DependencyFactoryResolver<D, P = any, E = any, C = any> = (params: EventDependencyResolverParams<P, E, C>) => D
```

Ill update the type in the test to better show that